### PR TITLE
Enhanced Order Validation

### PR DESCRIPTION
**Pull request by ScaredIbis**
# Motivation
- stricter initial validation to prevent false/invalid liquidity from being added to the OME orderbook
- Resolves https://mycelium-eth.atlassian.net/browse/LMDEVS-677

# Changes
- Add validation for order created timestamp
- Add validation for order expiry timestamp
- Add validation for order pair price crossing
- Add validation for user margin after performing order

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -82,14 +89,72 @@ app.post('/check', async (req, res) => {
...
+        });
+    }
+
+    // check if user has enough margin in the market
+    const perpetualSwapContract = new web3.eth.Contract(perpetualSwapABI, contractOrder.order.market)
+
+    const feeRate = await perpetualSwapContract.methods.feeRate().call()
+    const maxLeverage = await perpetualSwapContract.methods.maxLeverage().call()
+
+    const currentPosition = await perpetualSwapContract.methods.getBalance(contractOrder.order.maker).call()
```
Comment (raymogg) [inline index.js:131]: Might be premature optimisation, but we could cache these?

Early on we will only have 1 or 2 markets so loading in and reading these at each and every trade submission seems cumberson.

Probably scope creep on this ticket, but worth doing at some point in the future I'd say

**Review by raymogg — APPROVED** [review]

File: index.js (hunk 1 above)
Comment (ScaredIbis) [reply index.js:131]: its a good point, can `feeRate` and `maxLeverage` change once the contract is deployed?

Comment (ScaredIbis) [reply index.js:131]: @raymogg it looks like those two variables can be updated by the contract owner (not sure what that actually looks like in practice), In order to cache these effectively we would need a reliable way to invalidate the cache if these variables were to change on the contract.

I'll make a ticket to address the cache invalidation issue and proceed with this in its current state if you are cool with that?

Comment (ScaredIbis) [reply index.js:131]: ticketed up here: https://mycelium-eth.atlassian.net/browse/LMDEVS-712

**Pull request merged**